### PR TITLE
Log throughput Benchmark Scripts

### PR DESCRIPTION
**Pull request by jkosh44**
## Description
This PR adds python scripts that benchmark the log throughput of NoisePage. For the primary node, we define log throughput as how fast the database is generating log records (the cutoff being when the log record is serialized). For the replica node, we define log throughput as how fast the database is applying log records. We care about log throughput because under asynchronous replication log throughput acts as a good proxy for replication lag. If the primary node is generating logs at a faster rate than the replica node can apply them, then we'd expect the replica node to lag behind the primary node. The difference in log throughput gives us an approximate measure for that lag. Under synchronous replication, differences in log throughput between the primary and replica nodes can help explain increases in transaction latency compared to no replication. 

In order to capture log throughput, we utilize the existing metrics framework. Essentially we run a workload against the primary or replica node, then using the CSVs generated by the metrics framework we can get the total number of log records and the total time. Total log records divided by total time gives us log throughput.

For the replica node, we had to add metrics to the recovery manager. @mbutrovich @malin1993ml I know you said that recovery metrics weren't needed right now, but if they are actively intrusive then please let me know and I will find a workaround without adding recovery manager metrics. Alternatively, I could create a new category of metrics for the recovery manager instead of lumping it together with the logging metrics.

This PR builds off of #1534 and should wait for that to be merged first.

### Primary Node Log Throughput
To generate log records for the primary node we run the load phase of an OLTP Benchmark against the primary node. We then take the metrics produced by the Log Serializer Task

### Replica Node Log Throughput
To generate log records for the replica node we must first create a file filled with messages that a primary node would send to a replica node. There is a script called `log_scraper` that helps generate this file from the logs produced from turning messenger logging on in the database. We then use another script called `log_shipper` to send the messages to the replica node as fast as possible. We then take the metrics produced by the Recovery Manager.
The reason that we don't just use the primary node to generate messages is that we don't want the sending of messages to be slowed down by the actual query processing and execution. 

### Initial Results
Below are some initial results from running the script. All results are from running the load phase of YCSB with a scale factor of 3000 and 140 connection threads on Dev 10. The YCSB field size was manually changed to 1. For the replica's log throughput we used the first 20,000 messages generated from running this YCSB workload (increasing the number of messages didn't have an effect on log throughput but increased the run time of the script).

| Primary or Replica | Replication | Durability | Modifications | Log Throughput (records per millisecond) |
|--------------	|-----------	|--------------	|-------------	|-----------------------	|
| Primary | DISABLED | Async | None | 102.9403467 | 
| Primary | DISABLED | Sync | None | 98.8078754142502 |
| Primary | Sync | Sync | None | 1.13025120762569 |
| Primary | Async | Async | None | 1.235403819 |
| | | | | |
| Replica | N/A | Sync | None | 0.1226508677 |
| | | | | |
| Primary | Async | Async | Remove JSON serialization of messages | 88.39578983 |
| Primary | Async | Async | Remove message copying from serialization thread to messenger thread | 2.950203276 |
| Primary | Async | Async | Remove both JSON serialization and message copying | 92.43581384 |

## Performance
N/A

## Remaining Work
- [x] Move from Log Scraper to Log Sink

## Further Work
- Move JSON serialization from Log Serializer thread to Messenger thread to see if performance is improved.
- Investigate other message formats instead of JSON.
- Add Microbenchmark for serialization related tasks

### CODE REVIEW

**Comment by jkosh44** [comment]
Add functionality to automatically generate log record messages when testing log throughput for replica nodes. Builds off of #1557

**Comment by lmwnshn** [comment]
Adding blocked for dependence on #1534, but starting to review

**Comment by lmwnshn** [comment]
I can't comment on the file, 

- script/testing/replication/log_throughput/resources/log-messages-small.txt 

Did we come to a conclusion during our meeting on including this into the repo?

**Comment by jkosh44** [comment]
> I can't comment on the file,
> 
>     * script/testing/replication/log_throughput/resources/log-messages-small.txt
> 
> 
> Did we come to a conclusion during our meeting on including this into the repo?

I'll remove this

**Review by lmwnshn — COMMENTED** [review]
> Are num records and num txns being duplicated here?

Is the only genuine question, everything else is minor

File: src/storage/recovery/recovery_manager.cpp [hunk 1]
```diff
@@ -46,8 +46,37 @@ void RecoveryManager::WaitForRecoveryToFinish() {
...
+  // (by calling ComponentToRecord this increments the sample count)
+  bool logging_metrics_enabled =
+      common::thread_context.metrics_store_ != nullptr &&
+      common::thread_context.metrics_store_->ComponentToRecord(metrics::MetricsComponent::LOGGING);
+
   // Replay logs until the log provider no longer gives us logs
   while (true) {
+    // Record metrics from previous iteration
+    if (num_records > 0) {
+      if (common::thread_context.resource_tracker_.IsRunning()) {
```
[the remark this answers is not in the document]
Comment (jkosh44) [reply src/storage/recovery/recovery_manager.cpp:61]: I tried to copy the pattern used in LogSerializerTask and DiskLogConsumer

File: script/testing/replication/log_throughput/node_server.py [hunk 2]
```diff
@@ -0,0 +1,337 @@
...
+            raise RuntimeError("Unable to start database")
+        self.running = True
+        # Download and prepare OLTP Bench
+        self.oltp_server._clean_oltpbench()
+        self.oltp_server._download_oltpbench()
+        self.overwrite_ycsb_field_size()
+        self.oltp_server._build_oltpbench()
+
+    @staticmethod
+    def overwrite_ycsb_field_size():
```
[the remark this answers is not in the document]
Comment (jkosh44) [reply script/testing/replication/log_throughput/node_server.py:108]: Yeah, I think that'd be a good separate issue to work on. This particular function is actually even hackier than any of that. It's overwriting a variable in a Java file before compiling OLTPBench.

File: .gitignore [hunk 3]
```diff
@@ -277,6 +277,9 @@ wal.log
 **/oltp_output_*
 **/oltp_result
 
+# Log Throuput output
```
[the remark this answers is not in the document]
Comment (jkosh44) [reply .gitignore]: Fixed

File: src/storage/recovery/recovery_manager.cpp [hunk 4]
```diff
@@ -129,10 +166,13 @@ void RecoveryManager::ProcessCommittedTransaction(noisepage::transaction::timest
 
     if (IsSpecialCaseCatalogRecord(buffered_record)) {
       idx += ProcessSpecialCaseCatalogRecord(txn, &buffered_changes_map_[txn_id], idx);
+      records_processed++;
```
[the remark this answers is not in the document]
Comment (jkosh44) [reply src/storage/recovery/recovery_manager.cpp]: Moved

File: script/testing/replication/log_throughput/__main__.py [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+    aparser.add_argument("--oltp-benchmark",
+                         default=DEFAULT_BENCHMARK,
+                         choices=[YCSB, TPCC, TATP],
+                         help="Which OLTP benchmark to use")
+    aparser.add_argument("--oltp-scale-factor",
+                         default=DEFAULT_SCALE_FACTOR,
+                         help="Scale factor for OLTP benchmark")
+    aparser.add_argument("--log-file",
+                         help=f"File containing log record messages to send to replica node. If no file is specified "
+                              f"then logs will be generated using OLTP. only relevant when test_type is "
```
[the remark this answers is not in the document]
Comment (jkosh44) [reply script/testing/replication/log_throughput/__main__.py]: Fixed

File: script/testing/replication/log_throughput/log_scraper.py [hunk 6]
```diff
@@ -0,0 +1,69 @@
...
+    matches = re.findall(pattern, log)
+    if len(matches) != 1:
+        LOG.warn(f"Unknown log format: {log}")
+        return None
+    return matches[0]
+
+
+def main():
+    """
+    This file helps scrape logs produced by the Messenger component of NoisePage. It will go through the logs and
```
[the remark this answers is not in the document]
Comment (jkosh44) [reply script/testing/replication/log_throughput/log_scraper.py]: Fixed

File: script/testing/replication/log_throughput/node_server.py [hunk 7]
```diff
@@ -0,0 +1,337 @@
...
+        self.running = False
+        self.replica.delete_wal()
+
+    def is_running(self) -> bool:
+        return self.running
+
+
+class ImposterNode(NodeServer):
+    """
+    Python process that imitates a NoisePage server. This ii useful for mocking one side of replication
```
[the remark this answers is not in the document]
Comment (jkosh44) [reply script/testing/replication/log_throughput/node_server.py]: Fixed

File: script/testing/replication/log_throughput/metrics_file_util.py [hunk 8]
```diff
@@ -0,0 +1,72 @@
...
+"""
+This file exposes a bunch of helper methods for interacting with metric csv files
+"""
+
+
+def get_csv_file_path(file_name: str) -> str:
+    """
+    Get absolute path to csv metrics file
+
+    :param file_name Name of metrics file
```
[the remark this answers is not in the document]
Comment (jkosh44) [reply script/testing/replication/log_throughput/metrics_file_util.py]: I changed it anyway

File: script/testing/replication/log_throughput/__main__.py [hunk 9]
```diff
@@ -0,0 +1,65 @@
...
+                              f"{TestType.REPLICA.value}")
+    aparser.add_argument("--save-generated-log-file",
+                         default=False,
+                         action="store_true",
+                         help=f"If log record messages are generated then this flag causes those messages to be saved "
+                              f"to a file, only relevant when test_type is {TestType.REPLICA.value} and --log-file "
+                              f"param is not specified")
+    aparser.add_argument("--connection-threads",
+                         default=DEFAULT_CONNECTION_THREADS,
+                         help=f"Number of database connection threads to use, OLTP threads will scale accordingly")
```
[the remark this answers is not in the document]
Comment (jkosh44) [reply script/testing/replication/log_throughput/__main__.py]: Fixed

File: script/testing/replication/log_throughput/__main__.py [hunk 10]
```diff
@@ -0,0 +1,65 @@
...
+                         help=f"Whether or not async replication is enabled, only relevant when test_type is "
+                              f"{TestType.PRIMARY.value} and replication is enabled")
+    aparser.add_argument("--async-commit",
+                         default=False,
+                         action="store_true",
+                         help="Whether or not WAL async commit is enabled")
+    aparser.add_argument("--oltp-benchmark",
+                         default=DEFAULT_BENCHMARK,
+                         choices=[YCSB, TPCC, TATP],
+                         help="Which OLTP benchmark to use")
```
[the remark this answers is not in the document]
Comment (jkosh44) [reply script/testing/replication/log_throughput/__main__.py]: Fixed

File: script/testing/replication/log_throughput/__main__.py [hunk 11]
```diff
@@ -0,0 +1,65 @@
...
+    aparser.add_argument("--oltp-scale-factor",
+                         default=DEFAULT_SCALE_FACTOR,
+                         help="Scale factor for OLTP benchmark")
```
[the remark this answers is not in the document]
Comment (jkosh44) [reply script/testing/replication/log_throughput/__main__.py]: Fixed

File: script/testing/replication/log_throughput/constants.py [hunk 12]
```diff
@@ -0,0 +1,120 @@
...
+REPLICATION_ENABLED_KEY = "replication_enable"
+NETWORK_IDENTITY_KEY = "network_identity"
+ASYNC_REPLICATION_KEY = "async_replication_enable"
+WAL_ASYNC_COMMIT_KEY = "wal_async_commit_enable"
+METRICS_KEY = "metrics"
+USE_METRICS_THREAD_KEY = "use_metrics_thread"
+LOGGING_METRICS_ENABLED_KEY = "logging_metrics_enable"
+CONNECTION_THREAD_COUNT_KEY = "connection_thread_count"
+
+# Server Arg Default Values
```
[the remark this answers is not in the document]
Comment (jkosh44) [reply script/testing/replication/log_throughput/constants.py:17]: Added

File: src/include/metrics/logging_metric.h [hunk 13]
```diff
@@ -59,21 +63,27 @@ class LoggingMetricRawData : public AbstractRawData {
...
+  static constexpr std::array<std::string_view, 3> FILES = {"./log_serializer_task.csv", "./disk_log_consumer_task.csv",
+                                                            "./recovery_manager.csv"};
   /**
    * Columns to use for writing to CSV.
    * Note: This includes the columns for the input feature, but not the output (resource counters)
    */
-  static constexpr std::array<std::string_view, 2> FEATURE_COLUMNS = {"num_bytes, num_records, num_txns, interval",
-                                                                      "num_bytes, num_buffers, interval"};
+  static constexpr std::array<std::string_view, 3> FEATURE_COLUMNS = {
+      "num_bytes, num_records, num_txns, interval", "num_bytes, num_buffers, interval", "num_records, num_txns"};
```
[the remark this answers is not in the document]
Comment (jkosh44) [reply src/include/metrics/logging_metric.h:86]: Ok I finally remember what this is. It's an array of three strings. Each string contains the headers for a single CSV. So the first string "num_bytes, num_records, num_txns, interval" contains the headers for the log_serializer_task.csv. The second string "num_bytes, num_buffers, interval" contains the headers for the disk_log_consumer_task.csv. The third string "num_records, num_txns" contains the headers for recovery_manager.csv.

log_serializer_task.csv and recovery_manager.csv share some headers which is why it's duplicated.

**Comment by jkosh44** [comment]
@lmwnshn I responded to all your comments

**Review by lmwnshn — APPROVED** [review]
LGTM!

**Pull request merged**